lib/ddtrace/shim.rb

```diff
@@ -0,0 +1,112 @@
+module Datadog
+  # For wrapping methods and injecting hooks into other modules.
+  # Useful for adding instrumentation to libraries.
+  module Shim
+    # Creates and returns a Shim::Double.
+    # Pass a block to define methods to override.
+    def self.double(object, &block)
+      Double.new(object, &block)
+    end
+
+    def self.double?(object)
+      object.singleton_class.superclass <= Double
+    end
+
+    # Shorthands for wrapping methods
+    module Wrapping
+      def wrapped_methods
+        @wrapped_methods ||= Set.new
+      end
+
+      # Adds method block directly to the object.
+      # Block is evaluated in the context of the object.
+      # Faster than #wrap_method!
+      def inject_method!(name, &block)
+        define_singleton_method(name, &block).tap do
+          wrapped_methods.add(name)
+        end
+      end
+
+      # Adds method wrapper to the object.
+      # Block is evaluated in the original context of the block.
+      # Slower than #inject_method!
+      def wrap_method!(name, &block)
+        define_singleton_method(name) do |*original_args, &original_block|
+          block.call(*original_args, &original_block)
+        end
+        wrapped_methods.add(name)
+        name
+      end
+    end
+
+    # A "stand-in" that intercepts calls to another object. i.e. man-in-the-middle.
+    # This double forwards all methods to object, except those overriden.
+    # Useful if you want to intercept inbound behavior to an object without modifying
+    # the object in question, especially useful if the overridding behavior shouldn't be global.
+    class Double
+      extend Forwardable
+      include Shim::Wrapping
+
+      EXCLUDED_METHODS = [
+        :__binding__,
+        :__id__,
+        :__send__,
+        :define_singleton_method,
+        :extend,
+        :forward_methods!,
+        :itself,
+        :object_id,
+        :singleton_class,
+        :tap,
+        # From Shim::Wrapping
+        :wrapped_methods,
+        :inject_method!,
+        :wrap_method!
+      ].freeze
+
+      attr_reader :shim_target, :shim
+
+      # Pass this a block to override methods
+      def initialize(shim_target)
+        @shim = self
+        @shim_target = shim_target
+
+        # Wrap any methods
+        yield(self) if block_given?
+
+        # Forward methods
+        forwarded_methods = shim_target.public_methods - EXCLUDED_METHODS - wrapped_methods.to_a
+        forward_methods!(*forwarded_methods)
+      end
+
+      def forward_methods!(*forwarded_methods)
+        return if forwarded_methods.empty?
+
+        singleton_class.send(
+          :def_delegators,
```

Comment:
I tried looking up how `:def_delegators` is implemented. My concern is what is the performance impact of instrumenting object with many, many methods.

Reply:
As in if there are more arugments provided to that function than less? I wouldn't have imagined the number of arguments would matter here, but its a good question. I don't know how the performance cost scales, or how it compares to others, spare what I saw in https://gist.github.com/rbishop/7555357.